routers/api/v1/utils/git.go

```diff
@@ -27,10 +30,26 @@ func ResolveRefOrSha(ctx *context.APIContext, ref string) string {
 			return ""
 		}
 		if refSHA != "" {
-			return refSHA
+			sha = refSHA
+			break
 		}
 	}
-	return ref
+
+	if ctx.Repo.GitRepo != nil && ctx.Repo.GitRepo.LastCommitCache == nil {
+		commitsCount, err := cache.GetInt64(ctx.Repo.Repository.GetCommitsCountCacheKey(ref, true), func() (int64, error) {
+			commit, err := ctx.Repo.GitRepo.GetCommit(sha)
+			if err != nil {
+				return 0, err
+			}
+			return commit.CommitsCount()
+		})
+		if err != nil {
+			log.Error("Unable to get commits count for %s in %s. Error: %v", sha, ctx.Repo.Repository.FullName(), err)
```

Comment:
Should it return here? No need to create a cache if it's for some reason erroring.

Reply:
Honestly it depends on whether there is any way a commitscount could return an error but not mean that the repo is hosed.

I don't know. In any case the newcommitcache function will check the count returned. If the count is zero and the minimum count is zero then even if the repo is hosed it's still only a shim.

The lastcommitcache struct here isn't actually anything more than a lightweight shim for calling stuff. Handling and dealing with a broken repo can be deferred to all of the other code that is already written for dealing with that.